File: cpp/LeetCode202.cpp
#include <set>
#include <string>

class Solution {
public:
    bool isHappy(int n) {
        std::set<int> loop;
        while (true) {
            n = convert(n);
            if (n == 1) {
                return true;
            }
            if (loop.find(n) != loop.end()) {
                return false;
            } else {
                loop.insert(n);
            }
        }
    }

    int convert(int n) {
        std::string n_str = std::to_string(n);
        int sum = 0;
        for (int i = 0; i < n_str.length(); i++) {
            int digit = n_str[i] - '0';
            sum += std::pow(digit, 2);
        }
        return sum;
    }
};

File: python/LeetCode202.py
from functools import reduce

class Solution:
    def isHappy(self, n: int) -> bool:
        loop = {n}
        while True:
            n = reduce(lambda x, y: x + int(y)**2, str(n), 0)
            if n == 1:
                return True
            elif n in loop:
                return False
            else:
                loop.add(n)